src/Tools/ExternalAccess/Razor/Cohost/Handlers/Completion.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.LanguageServer;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.Completion;
+using Microsoft.CodeAnalysis.LanguageService;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+using LSP = Roslyn.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Razor.Cohost.Handlers;
+
+internal static class Completion
+{
+    private static CompletionListCache? s_completionListCache;
+
+    private static CompletionListCache GetCache()
+        => s_completionListCache ??= InterlockedOperations.Initialize(ref s_completionListCache, () => new());
+
+    public static async Task<LSP.CompletionList?> GetCompletionListAsync(
+        Document document,
+        LinePosition linePosition,
+        LSP.CompletionContext? completionContext,
+        bool supportsVSExtensions,
+        LSP.CompletionSetting completionCapabilities,
+        CancellationToken cancellationToken)
+    {
+        var cache = GetCache();
+
+        var position = await document
+            .GetPositionFromLinePositionAsync(linePosition, cancellationToken)
+            .ConfigureAwait(false);
+
+        var globalOptions = document.Project.Solution.Services.ExportProvider.GetService<IGlobalOptionService>();
+        var capabilityHelper = new CompletionCapabilityHelper(supportsVSExtensions, completionCapabilities);
+        var completionOptions = globalOptions.GetCompletionOptionsForLsp(document.Project.Language, capabilityHelper);
```

Comment:
I'm probably missing something simple - why can this not be put inside the shared method on the completion handler, passing in the global option service?

Reply:
same for the other global option usages

Reply:
It felt cleaner to make options an input, allowing the shared methods to perform the actual work. At the time, making the shared methods also gather up options seemed like making them do too much. Looking again, I think it's reasonable for them to just take an `IGlobalOptionService` and a `CompletionCapabilityHelper`.

Reply:
IIRC, I felt good about this until I noticed the `MaxCompletionListSize` option. :smile: